drivers/flash/flash_stm32wbx.c

```diff
@@ -97,17 +197,100 @@ static int erase_page(struct device *dev, uint32_t page)
 		return rc;
 	}
 
-	/* Check erase operation allowed */
-	if (regs->CR & FLASH_SR_PESD) {
-		return -EBUSY;
-	}
+	/* Implementation of STM32 AN5289, proposed in STM32WB Cube Application
+	 * BLE_RfWithFlash
+	 * https://github.com/STMicroelectronics/STM32CubeWB/tree/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_RfWithFlash
+	 */
+
+	do {
+		/**
+		 * When the PESD bit mechanism is used by CPU2 to protect its
+		 * timing, the PESD bit should be polled here.
+		 * If the PESD is set, the CPU1 will be stalled when reading
+		 * literals from an ISR that may occur after the flash
+		 * processing has been requested but suspended due to the PESD
+		 * bit.
+		 *
+		 * Note: This code is required only when the PESD mechanism is
+		 * used to protect the CPU2 timing.
+		 * However, keeping that code make it compatible with both
+		 * mechanisms.
+		 */
+		while (LL_FLASH_IsActiveFlag_OperationSuspended())
+			;
+
+		/* Enter critical section */
+		key = irq_lock();
+
+		/**
+		 *  Depending on the application implementation, in case a
+		 *  multitasking is possible with an OS, it should be checked
+		 *  here if another task in the application disallowed flash
+		 *  processing to protect some latency in critical code
+		 *  execution.
+		 *  When flash processing is ongoing, the CPU cannot access the
+		 *  flash anymore.Trying to access the flash during that time
+		 *  stalls the CPU.
+		 *  The only way for CPU1 to disallow flash processing is to
+		 *  take CFG_HW_BLOCK_FLASH_REQ_BY_CPU1_SEMID.
+		 */
+		cpu1_sem_status = LL_HSEM_GetStatus(HSEM,
+			CFG_HW_BLOCK_FLASH_REQ_BY_CPU1_SEMID);
+		if (cpu1_sem_status == 0) {
+			/**
+			 *  Check now if the CPU2 disallows flash processing to
+			 *  protect its timing. If the semaphore is locked, the
+			 *  CPU2 does not allow flash processing
+			 *
+			 *  Note: By default, the CPU2 uses the PESD mechanism
+			 *  to protect its timing, therefore, it is useless to
+			 *  get/release the semaphore.
+			 *
+			 *  However, keeping that code make it compatible with
+			 *  bothmechanisms.
+			 *  The protection by semaphore is enabled on CPU2 side
+			 *  with the command SHCI_C2_SetFlashActivityControl()
+			 *
+			 */
+			cpu2_sem_status = LL_HSEM_1StepLock(HSEM,
+				CFG_HW_BLOCK_FLASH_REQ_BY_CPU2_SEMID);
+			if (cpu2_sem_status == 0) {
+				/**
+				 * When CFG_HW_BLOCK_FLASH_REQ_BY_CPU2_SEMID is
+				 * taken, it is allowed to only erase one
+				 * sector.
+				 * When several sectors need to be erased,
+				 * the application shall first exit from the
+				 * critical section and try again.
+				 */
+				regs->CR |= FLASH_CR_PER;
+				regs->CR &= ~FLASH_CR_PNB_Msk;
+				regs->CR |= page << FLASH_CR_PNB_Pos;
+
+				regs->CR |= FLASH_CR_STRT;
```

Comment:
Does LL API offer any API call for doing that?

Reply:
No LL API for that.